[CSS] Tentative build fix for Ubuntu 2204 build bot

csaavedra opened the pull request:
#### ca99c1658f05b53b5378db4d030309a5da2c770f
<pre>
[CSS] Tentative build fix for Ubuntu 2204 build bot
<a href="https://bugs.webkit.org/show_bug.cgi?id=285598">https://bugs.webkit.org/show_bug.cgi?id=285598</a>

Reviewed by Sam Weinig.

The compiler complains about WTF::IsSmartPtr&lt;T&gt; being redefined.
Add a forward-declaration to prevent this and explicitly
add the namespace scope, as it seems that&apos;s what confuses
gcc.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:

Canonical link: <a href="https://commits.webkit.org/288716@main">https://commits.webkit.org/288716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6bc9664bd96d4d51236daea27b6c75d6324cf408

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84195 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89271 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35204 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86280 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11789 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65492 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23331 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2931 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45785 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2879 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30741 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34253 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77159 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31505 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90652 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83212 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11462 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8321 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73946 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72336 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73149 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18095 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17459 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2825 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11414 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16890 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/105631 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11263 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25510 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14739 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13036 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
```diff
@@ -189,6 +189,12 @@ template<Calc T> struct CSSValueChildrenVisitor<T> {
+template<typename T>
+struct IsSmartPtr;
+
+template<WebCore::CSS::Calc T> struct IsSmartPtr<T> {
```
I think a better fix for this would be to either:

a) Add the include for <wtf/GetPtr.h> to these files.
b) Add IsSmartPtr forward declaration to wtf/Forward.h and make sure wtf/Forward.h is included in these files.

csaavedra replied:
Thanks for the suggestions, you're right.

csaavedra replied:
Turns out the culprit is not having the namespace scope in the instantiation of the template, but I added the forward-declaration just for extra clarity as you suggested.

weinig replied:
Are both really needed? I would expect just the forward declaration would be enough. Either way, thanks!

csaavedra replied:
Yeah, somehow the scoped namespace is needed too. Thanks to you!

csaavedra committed:
[CSS] Tentative build fix for Ubuntu 2204 build bot
https://bugs.webkit.org/show_bug.cgi?id=285598

Reviewed by Sam Weinig.

The compiler complains about WTF::IsSmartPtr<T> being redefined.
Add a forward-declaration to prevent this and explicitly
add the namespace scope, as it seems that's what confuses
gcc.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:

Canonical link: https://commits.webkit.org/288716@main
